src/Simulation/QCTraceSimulator/DepthCounter.cs

```diff
@@ -244,20 +244,29 @@ public void OnRelease(object[] qubitsTraceData)
 
             if (optimizeDepth)
             {
-                // Doing width reuse heuristics and width computation.
+                // Performing width computation and applying reuse heuristics.
+                // First we fix busy time periods for all qubits involved.
                 foreach (QubitTimeMetrics q in qubitsTraceData.Cast<QubitTimeMetrics>())
                 {
-                    // If qubit wasn't used in any gate. We don't allocate it.
-                    if (q.EndTime.IsEqualTo(ComplexTime.Zero))
-                    {
-                        continue;
-                    }
-
                     // If qubit is not fixed in time we fix it at zero.
-                    if (q.StartTime.IsEqualTo(ComplexTime.MinValue))
-                    {
+                    if (q.StartTime.IsEqualTo(ComplexTime.MinValue)) {
                         q.StartTime = ComplexTime.Zero;
                     }
+                }
+
+                // Then we sort qubits ascending by the start of busy time. This would result in optimal reuse if:
+                // (1) We were not allowed to move gates in time.
+                // (2) We were sorting all qubits of the cirquit, not just this block.
+                // In reality these statements aren't true, so this is only a heuristic rather than the optimal reuse of qubits.
+                Array.Sort(qubitsTraceData, CompareAsQubitTimeMetricsByStartTime);
+
+                // Reuse qubits if possible or increase count of qubits used.
+                foreach (QubitTimeMetrics q in qubitsTraceData.Cast<QubitTimeMetrics>())
+                {
+                    // If qubit wasn't used in any gate. We don't allocate it.
+                    if (q.EndTime.IsEqualTo(ComplexTime.Zero)) {
```

Comment:
Is this correct even when there are gates right at the beginning that all have duration set to zero, and the qubit is only ever involved in these?

Reply:
Yes, this is correct. ComplexTime.Zero means not only zero depth, but also zero gates.

Reply:
Got it, thanks!